core/src/bytecode/value/mod.rs

```diff
@@ -32,30 +33,201 @@ pub type Array = Slice<NickelValue, 32>;
+}
+
+impl InlinePosIndex {
+    /// A special constant indicating that an inline value has no position set.
+    const NONE: InlinePosIndex = Self(u32::MAX);
+
+    /// Creates a position index from a usize, truncating higher bits if set.
+    pub fn from_usize_truncate(value: usize) -> Self {
+        Self(value as u32)
+    }
+}
+
+impl From<InlinePosIndex> for usize {
+    fn from(value: InlinePosIndex) -> Self {
+        value.0 as usize
+    }
+}
+
 /// The unified representation of Nickel values.
 ///
 /// A tagged pointer to a [reference-counted Nickel value block][ValueBlockRc], or an inline value.
 /// The two least significant bits of the pointer are used as the tag. See [ValueTag] for more
 /// details.
 #[derive(Debug)]
-pub struct NickelValue(usize);
+pub struct NickelValue {
+    data: usize,
+    // On 64-bits pointer-width archs, we can fit everything into one word. Otherwise, we stay safe
+    // and use a separate field for the position of inline values.
+    #[cfg(not(target_pointer_width = "64"))]
+    inline_pos_index: InlinePosIndex,
+}
 
+// pointer-width-specific implementation
+#[cfg(target_pointer_width = "64")]
 impl NickelValue {
-    /// The mask for the tag bits in a value pointer.
-    const VALUE_TAG_MASK: usize = 0b11;
+    /// The mask for the position indexbits in an inline value.
+    const INLINE_POS_MASK: usize = 0xFF_FF_FF_FF_00_00_00_00;
+
+    /// Returns `self` with the position index set to `pos` if `self` is an inline value, or
+    /// returns `self` unchanged otherwise.
+    pub fn with_inline_pos(mut self, pos: InlinePosIndex) -> Self {
+        if self.tag() == ValueTag::Inline {
+            self.data = (self.data & !Self::INLINE_POS_MASK) | (usize::from(pos) << 32);
+        }
+        self
+    }
 
-    /// Returns the tag bits of this value.
-    pub fn tag(&self) -> ValueTag {
-        (self.0 & Self::VALUE_TAG_MASK).try_into().unwrap()
+    /// Returns the position index of `self` is an inline value, or `None` otherwise.
+    pub fn inline_pos(&self) -> Option<InlinePosIndex> {
+        (self.tag() == ValueTag::Inline)
+            .then(|| InlinePosIndex::from_usize_truncate((self.data & Self::INLINE_POS_MASK) >> 32))
     }
 
-    /// Creates a new inline value.
+    /// Creates a new inline value with no position set.
     pub const fn inline(inline: InlineValue) -> Self {
-        // Safety: inline values are "pre-tagged", so they already have their low bits set to the
-        // Inline tag and are represented as `usize`
-        NickelValue(inline as usize)
+        NickelValue {
+            data: inline as usize | ((InlinePosIndex::NONE.0 as usize) << 32),
+        }
+    }
+
+    /// Converts `self` to an inline value bypassing any safety checks.
+    ///
+    /// # Safety
+    ///
+    /// `self` must hold a valid inline value.
+    pub unsafe fn as_inline_unchecked(&self) -> InlineValue {
+        // Safety: `InlineValue` is `#[repr(usize)]`, ensuring that it has the same layout as
```

Comment:
Maybe `InlineValue` should be `u32`?

Reply:
I suppose it would make sense now. I was going to argue that on exotic archs it could something else (like 16-bit usize), but 1. we probably don't really need to support anything else than 64 and 32 and 2. that would be dangerous because most likely we'll design inline values with a `u32` in mind but this might break in strange ways for untested archs other than 32/64. So better make clear that we expect 32 bits of information indeed.